`synapse` and `mssql` bugfixes and improvements

jorritsandbrink opened the pull request:
<!--
Thank you for submitting a pull request! Please provide a brief description of your changes below.
-->
### Description
This PR
- adds chunk separation handling for the `select_union` insert values writer type to fix #1173 
- sets the `max_rows_per_insert` property to prevent _"Some part of your SQL statement is nested too deeply. Rewrite the query or break it up into smaller queries."_ errors
- adds `mssql` and `synapse` to set of tested destinations in `test_insert_job_client.py` for better test coverage
- removes pipeline dependency in `SynapseStagingCopyJob.generate_sql` by passing the `job_client` as argument
- improves `PyOdbcMsSqlClient.rollback_transaction` such that it won't fail if there is nothing to rollback
- reorganizes logic in some places for better code quality

<!--
Please link any related issues. This helps us keep the PR focused and merge it faster.
-->
### Related Issues

- Fixes #1173

rudolfix reviewed (changes requested):
Please take a look at `test_insert_job_client`, you should add synapse there:
```
DEFAULT_SUBSET = ["duckdb", "redshift", "postgres"]
```
and also mssql (it also uses inserts)
you can then merge your test into it

jorritsandbrink commented:
@rudolfix I added `mssql` and `synapse` to `test_insert_job_client.py` and made some changes so the tests pass. Also removed the pipeline dependency to obtain a `job_client` in `SynapseStagingCopyJob.generate_sql` as discussed on Slack—this is done in an unsophisticated way that violates `from_table_chain` and `generate_sql` interfaces because the proper way would involve substantial code changes, type checking errors are ignored. See https://github.com/dlt-hub/dlt/pull/1174#issue-2220819706 for an overview of the different changes contained in this PR.

Can you review?

rudolfix reviewed (approved):
LGTM! thanks for improving the tests!!
